[Feat]: Gitlab p2 - let user add PAT via FE

malhotra5 opened the pull request:
- [ ] This change is worth documenting at https://docs.all-hands.dev/
- [ ] Include this change in the Release Notes. If checked, you **must** provide an **end-user friendly** description for your change below

**End-user friendly description of the problem this fixes or functionality that this introduces.**


---
**Give a summary of what the PR does, explaining any non-trivial design decisions.**


---
**Link of any specific issues this addresses.**

malhotra5 commented:
Note: I've modified the repo dropdown to work as follows

When only one provider is set, we show the current dropdown
![image](https://github.com/user-attachments/assets/5136cc14-9a5a-4232-9aa0-ce8bbbd452ac)


When two or more providers are set, we show a dropdown to choose the appropriate provider. After a provider is selected, they can interact with the repo dropdown as normal (scroll to see repos, type to search across repo in PAT and public ones)

1. This is a simpler mental model for users
2. When we allow multiple repo selection, its unlikely that users will select repos across different providers
3. Implementing this is easier. For example `search_repositories` doesn't need to be done across many providers, and its less complicated for the end user

<img width="798" alt="Screenshot 2025-03-21 at 4 50 51 PM" src="https://github.com/user-attachments/assets/768
<img width="857" alt="Screenshot 2025-03-21 at 4 51 06 PM" src="https://github.com/user-attachments/assets/0a2ec262-0c79-47dd-8a77-d15f7ee9be87" />
0478d-d71b-48ce-bda0-3c29c08bf444" />

rbren commented on frontend/src/routes/account-settings.tsx:
```diff
@@ -381,7 +390,7 @@ function AccountSettings() {
           <section className="flex flex-col gap-6">
             <h2 className="text-[28px] leading-8 tracking-[-0.02em] font-bold">
-              GitHub Settings
+              Provider Settings
```
```suggestion
              Git Provider Settings
```

malhotra5 commented:
Also, I've decided against renaming entire files for now. Mainly for review purposes, because it makes it hard to track the diff changes. Once I get a thumbs up I'll update files such as `frontend/src/components/features/github/github-repositories-suggestion-box.tsx` to `frontend/src/components/features/git/git-repositories-suggestion-box.tsx

malhotra5 commented on frontend/src/routes/account-settings.tsx:
```diff
@@ -436,16 +446,58 @@ function AccountSettings() {
+                  <b>
+                    {" "}
+                    <a
+                      href="https://gitlab.com/-/user_settings/personal_access_tokens?name=openhands-app&scopes=api,read_user,read_repository"
```
Ah nice catch

rbren replied:
Looks like we need it to create MRs unfortunately 🤷

rbren commented:
Dropdown still says GitHub. We should probably show GH and GL in separate sections, like we currently do for public/private

(btw no need to retrieve public GitLab repos IMO, not sure if you're doing that now)

![Uploading Screenshot 2025-03-21 at 5.08.47 PM.png…]()

malhotra5 commented:
Ah yes I've already implemented searching public GitLab repos (for consistency, it would be odd to search the ones for github but not gitlab imo)

I could update the dropdown place to show `Select GitLab project` instead when Gitlab is chosen as the provider

IMO its odd to intertwine multiple providers in the same dropdown (its just a tougher mental model for users, and hard to execute if the user has 100+ repos). Furthermore its unlikely that users will want to see both providers in the same dropdown too - usually you're working on a single repo, or even multiple repos for the same provider

rbren commented:
@malhotra5 I don't love the double dropdown. I think different sections, combined with the typeahead, will be better

> In the future if we allow multiple repo selection, its unlikely that users will select repos across different providers

Famous last words :P I do think we'll need to handle this, but it'll probably involve reworking this whole UI

rbren commented:
@malhotra5 what are the technical limitations that are blocking this?

FWIW there are so few public repos on GitLab that I don't think dropping those from the public list is a big deal

rbren commented:
I haven't gotten it to actually clone a gitlab repo. We probably need to start including the host here?
<img width="400" alt="Screenshot 2025-03-24 at 5 21 20 PM" src="https://github.com/user-attachments/assets/b286c09a-f09b-4257-9e9f-298861d4847e" />

malhotra5 commented:
I think pagination and ranking is are the main technical problems

200 github repos would drown out a few gitlab ones (especially since we arent able to rank by most recently pushed on github at least)

I'm OK putting them all in one dropdown, will rework this accordingly; we'll likely see multiple cursor pagination logic from the FE for this to work (which is OK but my hunch is can become cumbersome to maintain with 3+ providers - we'll see I'll take a first pass at it)

rbren commented:
Yeah I wonder if we need to just exhaust the repo list server-side, and return the full result (with some limit at say 1k repos). Not having things sorted properly is messing with me, and we're going to need the sorting for "Suggested Tasks"

malhotra5 commented:
<img width="657" alt="Screenshot 2025-03-25 at 1 34 09 PM" src="https://github.com/user-attachments/assets/4bea2631-016c-47bf-a609-f6d3e007ce1f" />
<img width="825" alt="Screenshot 2025-03-25 at 1 33 43 PM" src="https://github.com/user-attachments/assets/99122a5e-9ea1-40b3-8ea9-33780050a919" />

I've changed it so that repo pagination occurs in the backend and limits it to 1000 repositories per provider

I've also included search from every provider (as we only return a set limit of repos for this)

malhotra5 commented:
UPDATE: Thank you for pointing our the bug with cloning gitlab repos; I've fixed this now, and also added a microagent to make the agent aware that a GitLab token exists

<img width="765" alt="Screenshot 2025-03-25 at 3 19 22 PM" src="https://github.com/user-attachments/assets/d498fc0b-bd78-4588-b961-bef12816106e" />

rbren commented:
Something is still up with cloning. The repo gets cloned, but the files aren't there? All I can see is `.git`, and the agent is totally lost. GitHub seems to be working

A couple little nits on UI:

Can we put a loading spinner while the repo list loads? Or even just a message like `Loading...`
<img width="315" alt="Screenshot 2025-03-27 at 11 08 59 AM" src="https://github.com/user-attachments/assets/944c9e2d-680a-4dd6-a788-4a58dc4bf937" />

Is it easy to make `github` and `gitlab` part of the search keys for repos in the typeahead?

malhotra5 commented:
> Is it easy to make github and gitlab part of the search keys for repos in the typeahead?

Yes should be easyy

> Something is still up with cloning. The repo gets cloned, but the files aren't there? All I can see is .git, and the agent is totally lost. GitHub seems to be working

![image](https://github.com/user-attachments/assets/968f3647-e7b7-47d0-86a0-e08c70f082d2)

Have you pulled latest changes? I've added the microagent and fixed cloning with recent commits. Does this happen for private repos only or public repos as well?

> Can we put a loading spinner while the repo list loads? Or even just a message like Loading...

On it!

rbren commented:
@malhotra5 it's a private repo. Also noticing that the terminal appears to be trapped, so I'm guessing the `git clone` did something interactive...it takes a while too which suggests it's timing out silently
<img width="258" alt="Screenshot 2025-03-27 at 11 30 29 AM" src="https://github.com/user-attachments/assets/1e0e2c37-4387-4bcb-a3c0-6a0019e68075" />

malhotra5 commented:
@openhands please fix the failing frontend unit tests

malhotra5 commented:
Thank you for the feedback! I was able to repro and fix the cloning issue

I've also added the UI changes (provider based typeahead and spinner when repos are loading)

malhotra5 commented:
@openhands please fix the failing frontend tests
